### PR TITLE
Add mission scheduler

### PR DESCRIPTION
**Pull request by oysand**

### CODE REVIEW

File: backend/api/Services/ScheduledMissionService.cs [hunk 1]
```diff
@@ -17,21 +18,35 @@ public async Task<ScheduledMission> Create(ScheduledMission scheduledMission)
...
         }
 
+        public void Update(ScheduledMission scheduledMission)
+        {
+            _context.ScheduledMissions.Update(scheduledMission);
+            _context.SaveChanges();
+            RaiseScheduledMissionUpdatedEvent();
+        }
+
+        public async Task<List<ScheduledMission>> UpcomingPendingScheduledMissions()
```
Comment (GodVenn) [inline backend/api/Services/ScheduledMissionService.cs:44]: Renaming suggestion
```suggestion
        public async Task<List<ScheduledMission>> GetUpcomingPendingScheduledMissions()
```

**Review by GodVenn — CHANGES_REQUESTED** [review]
Please see comments

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                if (_upcomingScheduledMissions is not null)
+                {
+                    foreach (var upcomingScheduledMission in _upcomingScheduledMissions)
+                    {
+                        if (upcomingScheduledMission.Robot.Status is not RobotStatus.Available) { continue; }
+                        if (upcomingScheduledMission.StartTime > DateTimeOffset.UtcNow) { break; }
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:37]: They are ordered by time

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+                        bool startedSuccessfull = await StartScheduledMission(upcomingScheduledMission);
+                        if (startedSuccessfull) { UpdateUpcomingScheduledMissions(); }
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:40]: Agreed

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        private async Task<bool> StartScheduledMission(ScheduledMission scheduledMission)
+        {
+            try
+            {
+                var robot = await _robotService.Read(scheduledMission.Robot.Id);
+                if (robot == null) return false;
+                if (robot.Status is not RobotStatus.Available) return false;
+            }
+            catch (ArgumentNullException)
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:65]: Agreed, I think we should not tolerate this

File: backend/api/Services/ScheduledMissionService.cs (hunk 1 above)
Comment (oysand) [reply backend/api/Services/ScheduledMissionService.cs:44]: Agreed

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            _upcomingScheduledMissions = await _scheduledMissionService.UpcomingPendingScheduledMissions();
+        }
+
+        private async Task<bool> StartScheduledMission(ScheduledMission scheduledMission)
+        {
+            try
+            {
+                var robot = await _robotService.Read(scheduledMission.Robot.Id);
+                if (robot == null) return false;
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:62]: Agreed

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            UpdateUpcomingScheduledMissions();
+        }
+
+        private async void UpdateUpcomingScheduledMissions()
+        {
+            _upcomingScheduledMissions = await _scheduledMissionService.UpcomingPendingScheduledMissions();
+        }
+
+        private async Task<bool> StartScheduledMission(ScheduledMission scheduledMission)
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:58]: Agreed

File: backend/api/Services/ScheduledMissionService.cs [hunk 7]
```diff
@@ -17,21 +18,35 @@ public async Task<ScheduledMission> Create(ScheduledMission scheduledMission)
...
+        {
+            var scheduledMissions = await ReadAll();
+            var pendingScheduledMissions = scheduledMissions.SkipWhile(ev => !ev.Status.Equals(ScheduledMissionStatus.Pending));
+            return pendingScheduledMissions.OrderBy(ev => ev.StartTime).ToList();
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/Services/ScheduledMissionService.cs]: Agreed

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            var robot = await _robotService.Read(scheduledMission.Robot.Id);
+            if (robot is null)
+            {
+                _logger.LogInformation("Could not find robot {id}", scheduledMission.Robot.Id);
+                return false;
+            }
+            if (robot.Status is not RobotStatus.Available)
+            {
+                _logger.LogInformation("Robot {id} is not available", scheduledMission.Robot.Id);
```
Comment (GodVenn) [inline backend/api/EventHandlers/ScheduledMissionEventHandler.cs:68]:
```suggestion
                _logger.LogWarning("Robot {id} is not available", scheduledMission.Robot.Id);
```

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                return false;
+            }
+            try
+            {
+                var report = await _isarService.StartMission(robot: scheduledMission.Robot, missionId: scheduledMission.IsarMissionId);
+                _logger.LogInformation("Started mission {id}", scheduledMission.Id);
+            }
+            catch (MissionException e)
+            {
+                _logger.LogInformation(e, "Failed to start mission {id}", scheduledMission.Id);
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:78]: Agreed

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs (hunk 8 above)
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:68]: Agreed

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            _upcomingScheduledMissions = await _scheduledMissionService.GetUpcomingPendingScheduledMissions();
+        }
+
+        private async Task<bool> StartScheduledMission(ScheduledMission scheduledMission)
+        {
+            var robot = await _robotService.Read(scheduledMission.Robot.Id);
+            if (robot is null)
+            {
+                _logger.LogInformation("Could not find robot {id}", scheduledMission.Robot.Id);
```
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:63]: Agreed

File: backend/api/EventHandlers/ScheduledMissionEventHandler.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (oysand) [reply backend/api/EventHandlers/ScheduledMissionEventHandler.cs:40]: It was added below. Will update the log level to LogWarning

**Review by GodVenn — APPROVED** [review]
LGTM

**Pull request merged**